src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/CSharpTagHelperCodeRenderer.cs

```diff
@@ -340,18 +332,72 @@ private void RenderUnboundHTMLAttributes(IEnumerable<KeyValuePair<string, Chunk>
-                BuildBufferedWritingScope(children);
+                return;
             }
-            else
+
+            _writer.Write("if (")
+                   .Write(ExecutionContextVariableName)
+                   .Write(".")
+                   .Write(_tagHelperContext.ExecutionContextOutputPropertyName)
+                   .Write(".")
+                   .Write(_tagHelperContext.OutputContentSetPropertyName)
+                   .WriteLine(")");
+
+            // Render the body of the if statement, at this point in the codegen the TagHelperOutput's Content was set
+            // so we should just render its Content, no need to forcefully execute the child content.
+            using (_writer.BuildScope())
             {
-                // Render all of the tag helper children.
-                _bodyVisitor.Accept(children);
+                RenderTagOutput(_tagHelperContext.OutputGenerateContentMethodName);
+            }
+
+            _writer.Write("else if (")
+                   .Write(ExecutionContextVariableName)
+                   .Write(".")
+                   .Write(_tagHelperContext.ExecutionContextChildContentRetrievedPropertyName)
+                   .WriteLine(")");
+
+            // Render the body of the else if statement, at this point in the codegen the GetChildContentAsync method 
+            // was invoked but the TagHelperOutput's Content was not set. Call into GetChildContentAsync to retrieve
+            // the cached value of the content so we don't execute the child content twice.
+            using (_writer.BuildScope())
+            {
+                CSharpCodeVisitor.RenderPreWriteStart(_writer, _context);
+
+                _writer.WriteInstanceMethodInvocation(ExecutionContextVariableName,
+                                                      _tagHelperContext.ExecutionContextGetChildContentAsyncMethodName,
+                                                      endLine: false);
+
+                _writer.Write(".Result")
+                       .WriteEndMethodInvocation();
+            }
+
+            _writer.WriteLine("else");
+
+            // Render the body of the else statement, at this point in the codegen the GetChildContentAsync method 
+            // was not invoked and the TagHelperOutput's Content was not set. Call into ExecuteChildContentAsync to
+            // to execute and render child content.
+            using (_writer.BuildScope())
+            {
+                if (!string.IsNullOrEmpty(_context.TargetWriterName))
+                {
+                    _writer.WriteMethodInvocation(_tagHelperContext.StartWritingScopeMethodName, _context.TargetWriterName);
+                }
+
+                _writer.WriteInstanceMethodInvocation(
+                    ExecutionContextVariableName,
+                    _tagHelperContext.ExecutionContextExecuteChildContentAsyncMethodName,
+                    endLine: false);
+                _writer.WriteLine(".Wait();");
```

Comment:
nit: isn't this all within an `async Task` function?  if yes, can you instead gen `await` here (and elsewhere)?